### PR TITLE
Added comprehensive unit tests for datetime utilities and type estimation functions. 

### PR DESCRIPTION
**Pull request by UnitaryIron**
# These were the major changes:

- Added timezone awareness tests - Ensures both aware/naive datetime objects are coerced to UTC.

- Integrated Hypothesis property-based testing - Tests wide range of datetime inputs with edge cases.

- Added ISO 8601 format validation - Verifies common API datetime string formats work correctly.

- Added calendar boundary tests - Tests leap years, year-end, and epoch boundaries.

- Added idempotency test - Ensures parsing same object twice returns identical result.

- Fixed tuple return type - Changed from returning [] to () for tuple type estimation.

- Improved error message validation - Added check for "format" in DateTimeError message.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -56,6 +54,9 @@ pydantic-settings = ">=2.0.0"
 [tool.poetry.group.dev.dependencies]
 black = "^26.1.0"
 uvicorn = "^0.40.0"
+async-asgi-testclient = "^1.4.11"
+fakeredis = ">=2.33.0"
```
Comment (mdconaway) [inline pyproject.toml]: `FakeRedis` is actually used by the framework supplied redis adapter here: https://github.com/mdconaway/fastapi-cruddy-framework/blob/master/fastapi_cruddy_framework/adapters.py#L218

(An app developer can opt into FakeRedis via adapter arguments if they choose to run their application with a single worker and want to maintain the logical paradigm of redis existing for session storage, etc)

So this dependency should remain in the `tool.poetry.dependencies` section

File: pyproject.toml [hunk 2]
```diff
@@ -56,6 +54,9 @@ pydantic-settings = ">=2.0.0"
 [tool.poetry.group.dev.dependencies]
 black = "^26.1.0"
 uvicorn = "^0.40.0"
+async-asgi-testclient = "^1.4.11"
```
Comment (mdconaway) [inline pyproject.toml]: `async-asgi-testclient` is incorporated into some of the testing components the cruddy framework exports to app developers to give them a batteries included development experience. Some effort should probably be put into splitting these things out into different modules at some point like `cruddy core` vs `cruddy testing components`, but for now we probably need to leave this test client in `tool.poetry.dependences`.

For additional info, you can see the test helpers module:

https://github.com/mdconaway/fastapi-cruddy-framework/blob/master/fastapi_cruddy_framework/test_helpers.py#L3

**Comment by mdconaway** [comment]
Overall looks good. I think two lines of the pyproject.toml changes need to be reverted, but once that is done we can get this merged in. The estimate_simple_example can probably also be updated to do something like:
```
assert isinstance(estimate_simple_example(Union[None, tuple]), list)
```
per the comment below.

File: tests/unit/test_utils.py [hunk 3]
```diff
@@ -118,9 +179,9 @@ async def test_estimate_complex_list_of_dict():
 
 @mark.dependency()
 async def test_estimate_tuple():
-    assert estimate_simple_example(tuple) == []
-    assert estimate_simple_example(Union[None, tuple]) == []
-    assert estimate_simple_example(Optional[tuple]) == []
+    assert estimate_simple_example(tuple) == ()
```
Comment (mdconaway) [inline tests/unit/test_utils.py]: So this is an interesting case where the schema docs example function needs to convert a python type into something that can be represented in the server's openapi.json. Since tuples are python only, and cannot be represented in JSON form, this estimator function must coerce any tuple typing it discovers into a list so that it can be represented in JSON for swagger docs, etc. Even though the input type to this function is a tuple in this test, we are actually expecting it to return a list when we auto-generate schema examples if they are not provided by an app developer.

We may actually want to do something like:
```
assert isinstance(estimate_simple_example(Union[None, tuple]), list)
```

To confirm we are actually converting tuple field types, or annotated fields to a list for JSON representation.

Comment (mdconaway) [reply tests/unit/test_utils.py]: For reference, see: https://github.com/mdconaway/fastapi-cruddy-framework/blob/master/fastapi_cruddy_framework/util.py#L254

**Comment by UnitaryIron** [comment]
Thanks for the detailed review. I understand what you said about JSON not supporting tuples. I've changed it into:
```python
assert isinstance(estimate_simple_example(tuple), list)
```
Also, the dependencies of `fakeredis` and `async asgi test client` have been moved from `dev dependencies` to `tool.poetry.dependencies`.

**Comment by mdconaway** [comment]
So with the latest changes on this PR, I pulled the latest, ran `poetry lock` and `poetry install`, and then `make test`.

I also reset my dependencies and noticed that in python 3.13, we need to add `greenlet` as a dependency for some operating systems.

Please add this line as a project dependency to `pyproject.toml`

```
greenlet = "*"
```

Here are the test results I got:
```
fastapi-cruddy-framework % make test
Running test
poetry run coverage run --source=fastapi_cruddy_framework -m pytest tests && poetry run coverage report -m && poetry run coverage xml --fail-under 80
=========================================================================================================================================================== test session starts ===========================================================================================================================================================
platform darwin -- Python 3.13.12, pytest-9.0.2, pluggy-1.5.0
configfile: pytest.ini
plugins: anyio-4.9.0, dependency-0.6.0, typeguard-4.4.2, hypothesis-6.152.4
collected 101 items                                                                                                                                                                                                                                                                                                                       

tests/actions/test_action_reuse.py ...                                                                                                                                                                                                                                                                                              [  2%]
tests/crud/test_composite_keys.py ....                                                                                                                                                                                                                                                                                              [  6%]
tests/crud/test_create.py .....                                                                                                                                                                                                                                                                                                     [ 11%]
tests/crud/test_delete.py ....                                                                                                                                                                                                                                                                                                      [ 15%]
tests/crud/test_graphql.py ...                                                                                                                                                                                                                                                                                                      [ 18%]
tests/crud/test_read_many.py .............                                                                                                                                                                                                                                                                                          [ 31%]
tests/crud/test_read_one.py ........                                                                                                                                                                                                                                                                                                [ 39%]
tests/crud/test_relationships.py ................                                                                                                                                                                                                                                                                                   [ 55%]
tests/crud/test_update.py .....                                                                                                                                                                                                                                                                                                     [ 60%]
tests/policies/test_fundamentals.py ...                                                                                                                                                                                                                                                                                             [ 63%]
tests/query_forge/test_casting.py .                                                                                                                                                                                                                                                                                                 [ 64%]
tests/unit/test_utils.py F......F....................                                                                                                                                                                                                                                                                               [ 92%]
tests/websockets/test_broadcasts.py .                                                                                                                                                                                                                                                                                               [ 93%]
tests/websockets/test_direct_messages.py .                                                                                                                                                                                                                                                                                          [ 94%]
tests/websockets/test_kill_commands.py ....                                                                                                                                                                                                                                                                                         [ 98%]
tests/websockets/test_rooms.py ..                                                                                                                                                                                                                                                                                                   [100%]

================================================================================================================================================================ FAILURES =================================================================================================================================================================
________________________________________________________________________________________________________________________________________________________ test_utc_datetime_parser _________________________________________________________________________________________________________________________________________________________

    @mark.dependency()
    async def test_utc_datetime_parser():
        assert (
            type(
                parse_and_coerce_to_utc_datetime(
                    datetime(year=2012, month=2, day=1, hour=12, minute=46)
                )
            )
            is datetime
        )
        assert (
            type(
                parse_and_coerce_to_utc_datetime(
                    str(datetime(year=2012, month=2, day=1, hour=12, minute=46))
                )
            )
            is datetime
        )
    
        with raises(DateTimeError) as exc_info:
            parse_and_coerce_to_utc_datetime("garbage")
    
>       assert "format" in str(exc_info.value)
E       AssertionError: assert 'format' in 'Invalid datetime'
E        +  where 'Invalid datetime' = str(DateTimeError('Invalid datetime'))
E        +    where DateTimeError('Invalid datetime') = <ExceptionInfo DateTimeError('Invalid datetime') tblen=3>.value

tests/unit/test_utils.py:39: AssertionError
---------------------------------------------------------------------------------------------------------------------------------------------------------- Captured stdout setup ----------------------------------------------------------------------------------------------------------------------------------------------------------
2026-05-08 12:29:38,923 loglevel=INFO   logger=examples.fastapi_cruddy_sqlite.middleware.requestlogger dispatch() L20   rid=18PSXX start request path=/users/authorization
2026-05-08 12:29:38,923 loglevel=INFO   logger=examples.fastapi_cruddy_sqlite.middleware.requestlogger dispatch() L25   rid=18PSXX completed_in=0.32ms status_code=200
2026-05-08 12:29:38,924 loglevel=INFO   logger=examples.fastapi_cruddy_sqlite.middleware.requestlogger dispatch() L20   rid=9AY4F8 start request path=/users/authorization
2026-05-08 12:29:38,924 loglevel=INFO   logger=examples.fastapi_cruddy_sqlite.middleware.requestlogger dispatch() L25   rid=9AY4F8 completed_in=0.29ms status_code=200
----------------------------------------------------------------------------------------------------------------------------------------------------------- Captured log setup ------------------------------------------------------------------------------------------------------------------------------------------------------------
INFO     examples.fastapi_cruddy_sqlite.middleware.requestlogger:requestlogger.py:20 rid=18PSXX start request path=/users/authorization
INFO     examples.fastapi_cruddy_sqlite.middleware.requestlogger:requestlogger.py:25 rid=18PSXX completed_in=0.32ms status_code=200
INFO     examples.fastapi_cruddy_sqlite.middleware.requestlogger:requestlogger.py:20 rid=9AY4F8 start request path=/users/authorization
INFO     examples.fastapi_cruddy_sqlite.middleware.requestlogger:requestlogger.py:25 rid=9AY4F8 completed_in=0.29ms status_code=200
_______________________________________________________________________________________________________________________________________________ test_iso_string_variants[20120201T124600Z] ________________________________________________________________________________________________________________________________________________

iso_string = '20120201T124600Z'

    @mark.parametrize("iso_string", [
        "2012-02-01T12:46:00Z",
        "2012-02-01 12:46:00+00:00",
        "2012-02-01T12:46:00.000123Z",
        "20120201T124600Z",
    ])
    async def test_iso_string_variants(iso_string):
>       result = parse_and_coerce_to_utc_datetime(iso_string)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

tests/unit/test_utils.py:80: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
fastapi_cruddy_framework/util.py:179: in parse_and_coerce_to_utc_datetime
    return coerce_to_utc_datetime(parse_datetime(value))
                                  ^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

value = '20120201T124600Z'

    def parse_datetime(value: datetime | str | bytes | int | float) -> datetime:
        """
        Parse a datetime/int/float/string and return a datetime.datetime.
    
        This function supports time zone offsets. When the input contains one,
        the output uses a timezone with a fixed offset from UTC.
    
        Raise ValueError if the input is well formatted but not a valid datetime.
        Raise ValueError if the input isn't well formatted.
        """
        if isinstance(value, datetime):
            return value
    
        number = get_numeric(value, "datetime")
        if number is not None:
            return from_unix_seconds(number)
    
        if isinstance(value, bytes):
            value = value.decode()
    
        match: Match = datetime_re.match(value)  # type: ignore
        if match is None:
>           raise DateTimeError("Invalid datetime", code=None)
E           fastapi_cruddy_framework.util.DateTimeError: Invalid datetime

fastapi_cruddy_framework/util.py:140: DateTimeError
========================================================================================================================================================= short test summary info =========================================================================================================================================================
FAILED tests/unit/test_utils.py::test_utc_datetime_parser - AssertionError: assert 'format' in 'Invalid datetime'
FAILED tests/unit/test_utils.py::test_iso_string_variants[20120201T124600Z] - fastapi_cruddy_framework.util.DateTimeError: Invalid datetime
====================================================================================================================================================== 2 failed, 99 passed in 4.82s =======================================================================================================================================================
```

Please either:
1. Update the tests to pass
or
2. Update the framework code to pass the tests you have introduced.

Thanks!

**Comment by UnitaryIron** [comment]
## I've updated the pyproject.toml to include greenlet, and fixed test_utils.py.

### First error:
`FAILED tests/unit/test_utils.py::test_utc_datetime_parser - AssertionError: assert 'format' in 'Invalid datetime'`

In the test, a string `"garbage"` is passed to `parse_and_coerce_to_utc_datetime()`. It was expected to raise a `DateTimeError` containing the word "format" in the error message. But the actual error message being returned was `"Invalid datetime"`. As the word `"format"` isn't present in `"Invalid datetime"` the assertion fails. 

### Fix:
This was fixed by updating the test. The assertion was changed to look for the actual string returned:
`assert "Invalid datetime" in str(exc_info.value)`
### Second error:
`FAILED tests/unit/test_utils.py::test_iso_string_variants[20120201T124600Z] - fastapi_cruddy_framework.util.DateTimeError: Invalid datetime`

This test uses `@mark.parametrize` to check many ISO datetime formats. It passed for the standard formats but it failed for this specific format - `"20120201T124600Z"` The traceback showed that the regex `(datetime_re.match(value))` at line 138 of `util.py` does not recognize this "compact" string format as a valid datetime, so it raises a `DateTimeError`.

### Fix:
To align the test with the current framework's capabilities, it was removed from the list of parameters from the test file. 

> **As I couldn't run the tests locally, please let me know if it passed at your end.**

**Review by mdconaway — APPROVED** [review]
LGTM
